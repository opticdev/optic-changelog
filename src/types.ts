export type Changelog = {
  data: {
    endpoints: {
      change: {
        category: string
      }
      path: string
      method: string
    }[]
  }
}

export type JobInputs = {
  repoToken?: string
  subscribers: string[]
  opticSpecPath: string
}

export type RepoInfo = {
  owner: string
  repo: string
  prNumber?: number
  headSha: string
}

export type PrComment = {
  id: number
  body: string
}

export type PrInfo = {
  baseSha: string
  baseBranch: string
}

export declare interface IGitProvider {
  getFileContent(sha: string, path: string): Promise<string>
  getPrInfo(prNumber: number): Promise<PrInfo>
  updatePrComment(commentId: number, body: string): Promise<void>
  createPrComment(prNumber: number, body: string): Promise<void>
  getPrBotComments(prNumber: number): Promise<PrComment[]>
}

export declare interface IJobRunner {
  debug(message: string): void
  info(message: string): void
  setFailed(message: string | Error): void
}
