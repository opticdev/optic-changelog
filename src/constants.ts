export const API_BASE = process.env.OPTIC_API_BASE || 'https://api.useoptic.com'
export const CLOUD_SPEC_VIEWER_BASE =
  process.env.CLOUD_SPEC_VIEWER_BASE || 'https://spec.useoptic.com'
export const COMMENT_HEADER_IMG =
  'https://raw.github.com/opticdev/optic/develop/assets/changelog_header.png'
export const SENTRY_DSN =
  'https://b5d9462fe8e741c1a01c1f0ec8562241@o446328.ingest.sentry.io/5741686'
