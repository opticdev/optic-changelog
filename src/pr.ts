import {Changelog} from './types'

const REGEX = /\n\n<!-- optic = (.*) -->/

export function isOpticComment(body: string): boolean {
  return body.match(REGEX) ? true : false
}

export function getMetadata(body: string): any {
  const match = body.match(REGEX)

  if (match) {
    return JSON.parse(match[1])
  }

  return {}
}

export function setMetadata(body: string, data: any): string {
  let currentData = {}

  const bodyText = body.replace(REGEX, (_, json) => {
    currentData = JSON.parse(json)
    return ''
  })

  return `${bodyText}\n\n<!-- optic = ${JSON.stringify({
    ...currentData,
    ...data
  })} -->`
}

export function generateCommentBody(
  changes: Changelog,
  subscribers: string[]
): string {
  const results = {
    added: 0,
    updated: 0,
    removed: 0
  }

  for (const endpoint of changes.data.endpoints) {
    switch (endpoint.change.category) {
      case 'added':
        results.added++
        break
      case 'updated':
        results.updated++
        break
      case 'removed':
        results.removed++
        break
    }
  }

  const timestamp = new Date()
    .toISOString()
    .replace(/T/, ' ')
    .replace(/\..+/, '')

  const baseBody = `## Optic Changelog
  
* Endpoints added: ${results.added}
* Endpoints updated: ${results.updated}

Last updated: ${timestamp}

[View documentation](${changes.data.opticUrl})
`

  if (subscribers.length) {
    const subscriberText = subscribers
      .map(subscriber => `@${subscriber}`)
      .join(', ')
    return `${baseBody}
---

Pinging subscribers ${subscriberText}`
  }

  return baseBody
}
