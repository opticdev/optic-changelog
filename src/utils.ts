import * as Sentry from '@sentry/node'
import {SpanContext} from '@sentry/types/dist'
import {Span} from '@sentry/tracing'

export async function sentryInstrument<T>(
  spanParams: Partial<SpanContext>,
  cb: (
    transaction: ReturnType<typeof Sentry['startTransaction']>,
    span: Span
  ) => Promise<T>
): Promise<T> {
  const existingTxn = !!Sentry.getCurrentHub().getScope()?.getTransaction()
  const transaction =
    Sentry.getCurrentHub().getScope()?.getTransaction() ??
    Sentry.startTransaction({
      op: 'gitbot_run',
      name: 'Gitbot Run'
    })

  const existingSpan = Sentry.getCurrentHub().getScope()?.getSpan()

  const span = existingSpan
    ? existingSpan.startChild(spanParams)
    : transaction.startChild(spanParams)

  Sentry.getCurrentHub().configureScope(s => s.setSpan(span))
  try {
    return await cb(transaction, span)
  } catch (e) {
    Sentry.captureException(e)
    throw e
  } finally {
    span.finish()
    if (existingSpan) {
      Sentry.getCurrentHub().configureScope(s => s.setSpan(existingSpan))
    } else {
      Sentry.getCurrentHub().configureScope(s => s.setSpan(transaction))
    }
    if (!existingTxn) {
      // We started the txn, so we need to finish it
      transaction.finish()
    }
  }
}
