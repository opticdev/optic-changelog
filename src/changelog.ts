import {Changelog} from './types'
import {setMetadata, isOpticComment, generateCommentBody} from './pr'

export async function runOpticChangelog({
  subscribers,
  opticSpecPath,
  gitProvider,
  headSha,
  baseSha,
  baseBranch,
  prNumber,
  jobRunner
}): Promise<void> {
  let headContent: string, baseContent: string

  // This may fail if the file was moved or it's a new Optic setup
  try {
    headContent = await gitProvider.getFileContent(headSha, opticSpecPath)
  } catch (error) {
    // Failing silently here
    jobRunner.info(
      `Could not find the Optic spec in the current branch. Looking in ${opticSpecPath}.`
    )
    return
  }

  // This may fail if the file was moved or it's a new Optic setup
  try {
    baseContent = await gitProvider.getFileContent(baseSha, opticSpecPath)
  } catch (error) {
    // Failing silently here
    jobRunner.info(
      `Could not find the Optic spec in the base branch ${baseBranch}. Looking in ${opticSpecPath}.`
    )
    return
  }

  // TODO: use new changelog library here
  const changes = getChangelogData({
    from: JSON.parse(baseContent),
    to: JSON.parse(headContent)
  })

  if (changes.data.endpoints.length === 0) {
    jobRunner.info('No API changes in this PR.')
    return
  }

  const message = generateCommentBody(changes, subscribers)
  const body = setMetadata(message, {})

  jobRunner.debug('Created body for comment')
  jobRunner.debug(body)

  try {
    // TODO: probably should be simplified a bit
    const existingBotComments = (
      await gitProvider.getPrBotComments(prNumber)
    ).filter(comment => isOpticComment(comment.body!))

    if (existingBotComments.length) {
      const comment = existingBotComments[0]
      jobRunner.debug(`Updating comment ${comment.id}`)
      // TODO: need to pull out metadata and combine with new (maybe)
      await gitProvider.updatePrComment(comment.id, body)
    } else {
      jobRunner.debug(`Creating comment for PR ${prNumber}`)
      await gitProvider.createPrComment(prNumber, body)
    }
  } catch (error) {
    jobRunner.info(
      `There was an error creating or updating a PR comment. Error message: ${error.message}`
    )
    return
  }
}

// TODO: this is fake data for now
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function getChangelogData(options: object): Changelog {
  return {
    data: {
      opticUrl: 'https://example.com',
      endpoints: [
        {
          change: {
            category: 'added'
          },
          path: '/foo',
          method: 'get'
        },
        {
          change: {
            category: 'updated'
          },
          path: '/bar',
          method: 'post'
        }
      ]
    }
  }
}
