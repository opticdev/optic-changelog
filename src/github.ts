import * as core from '@actions/core'
import * as github from '@actions/github'
import {GitHub} from '@actions/github/lib/utils'
import {IGitProvider, JobInputs, RepoInfo, PrInfo, PrComment} from './types'

export function getJobInputs(): JobInputs {
  const repoToken = core.getInput('GITHUB_TOKEN') || process.env['GITHUB_TOKEN']

  const subscribers = core
    .getInput('subscribers')
    .split(',')
    .map(subscriber => subscriber.trim())

  const opticSpecPath = core.getInput('OPTIC_SPEC_PATH')
  const opticApiKey = core.getInput('OPTIC_API_KEY')

  return {repoToken, subscribers, opticSpecPath, opticApiKey}
}

export function getRepoInfo(): RepoInfo {
  const {
    payload: {repository, pull_request: pullRequest},
    sha: headSha
  } = github.context
  const {full_name: repoFullName = ''} = repository ?? {}
  const [owner, repo] = repoFullName.split('/')

  return {
    prNumber: pullRequest?.number,
    owner,
    repo,
    headSha
  }
}

export class GitHubRepository implements IGitProvider {
  constructor(
    private octokit: InstanceType<typeof GitHub>,
    public owner: string,
    public repo: string
  ) {}

  async getFileContent(sha: string, path: string): Promise<string> {
    const response = await this.octokit.repos.getContent({
      owner: this.owner,
      repo: this.repo,
      path,
      ref: sha
    })

    if (!('content' in response.data)) return ''

    const buff = Buffer.from(response.data.content, 'base64')
    return buff.toString('utf-8')
  }

  getRepoInfo(): RepoInfo {
    return getRepoInfo()
  }

  async getPrInfo(prNumber: number): Promise<PrInfo> {
    const prInfo = await this.octokit.pulls.get({
      owner: this.owner,
      repo: this.repo,
      pull_number: prNumber
    })
    return {
      baseSha: prInfo.data.base.sha,
      baseBranch: prInfo.data.base.ref
    }
  }

  async updatePrComment(commentId: number, body: string): Promise<void> {
    await this.octokit.issues.updateComment({
      owner: this.owner,
      repo: this.repo,
      comment_id: commentId,
      body
    })
  }

  async createPrComment(prNumber: number, body: string): Promise<void> {
    await this.octokit.issues.createComment({
      owner: this.owner,
      repo: this.repo,
      issue_number: prNumber,
      body
    })
  }

  async getPrBotComments(prNumber: number): Promise<PrComment[]> {
    const issueComments = await this.octokit.issues.listComments({
      owner: this.owner,
      repo: this.repo,
      issue_number: prNumber
    })

    const existingBotComments = issueComments.data.filter(
      comment => comment.user?.login === 'github-actions[bot]'
    )

    return existingBotComments.map(comment => ({
      id: comment.id,
      body: comment.body!
    }))
  }
}
