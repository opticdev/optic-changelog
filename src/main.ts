import * as core from '@actions/core'
import * as github from '@actions/github'
import {runOpticChangelog} from './changelog'
import {getJobInputs, getRepoInfo, GitHubRepository} from './github'

async function run(): Promise<void> {
  try {
    const {repoToken, subscribers, opticSpecPath} = getJobInputs()

    if (!repoToken) {
      throw new Error(
        'Please provide a GitHub token. Set one with the repo-token input or GITHUB_TOKEN env variable.'
      )
    }

    const octokit = github.getOctokit(repoToken)

    const {prNumber, owner, repo, headSha} = getRepoInfo()

    // We exit quietly when it's not a pull request
    if (!prNumber) {
      core.info('Not a pull request')
      return
    }

    const gitHubRepo = new GitHubRepository(octokit, owner, repo)
    const {baseSha, baseBranch} = await gitHubRepo.getPrInfo(prNumber)

    await runOpticChangelog({
      subscribers,
      opticSpecPath,
      gitHubRepo,
      headSha,
      baseBranch,
      baseSha,
      prNumber
    })
  } catch (error) {
    core.setFailed(error.message)
  }
}

// Don't auto-execute in the test environment
if (process.env['NODE_ENV'] !== 'test') {
  run()
}
