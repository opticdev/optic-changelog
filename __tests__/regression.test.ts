import * as fs from 'fs'
import * as path from 'path'
import {generateEndpointChanges} from '@useoptic/changelog'
import {ChangelogParams, runOpticChangelog} from '../src/changelog'
import {IGitProvider} from '../src/types'

const BASE_SHA = 'base-sha'
const UNRELATED_SHA = 'unrelated-sha'
const REGRESSION_1 = 'regression-1'

const pathMap = {
  [BASE_SHA]: null,
  [UNRELATED_SHA]: 'specs/add-method/next.json',
  [REGRESSION_1]: 'specs/regression/regression_1.json'
}

const baseGitProvider: IGitProvider = {
  getFileContent: async (sha: string) => {
    if (!sha) {
      return '[]'
    } else {
      return fs.readFileSync(path.join(__dirname, pathMap[sha])).toString()
    }
  },
  getPrBotComments: async () => [],
  updatePrComment: jest.fn(),
  createPrComment: jest.fn(),
  getPrInfo: jest.fn(),
  getRepoInfo: jest.fn()
}

const mockJobRunner = {
  info: jest.fn(),
  warning: jest.fn(),
  debug: jest.fn(),
  setFailed: jest.fn(),
  exportVariable: jest.fn()
}

const baseOpticChangelog: ChangelogParams = {
  apiKey: 'pizza',
  subscribers: [],
  opticSpecPath: '.optic/api/specification.yml',
  gitProvider: baseGitProvider,
  headSha: '',
  baseSha: BASE_SHA,
  baseBranch: 'main',
  prNumber: 100,
  jobRunner: mockJobRunner,
  generateEndpointChanges,
  uploadSpec: jest
    .fn()
    .mockResolvedValue({specId: 'spec-id', personId: 'person-id'})
}

describe('Regression', () => {
  beforeAll(() => {
    Date.prototype.toISOString = jest.fn(() => '2021-04-21T15:06:33.601Z')
  })

  it('shouldnt error on unrelated history', async () => {
    await runOpticChangelog({
      ...baseOpticChangelog,
      baseSha: UNRELATED_SHA,
      headSha: REGRESSION_1
    })
    expectCommentWasCreated()
    expect(mockJobRunner.exportVariable).toBeCalledTimes(1)
  })
})

function expectCommentWasCreated() {
  expect(baseGitProvider.createPrComment).toBeCalledTimes(1)
  expect(baseGitProvider.updatePrComment).toBeCalledTimes(0)
  expect(mockJobRunner.setFailed).toBeCalledTimes(0)
  expect(mockJobRunner.debug.mock.calls).toMatchSnapshot()
  expect(mockJobRunner.warning.mock.calls).toMatchSnapshot()
}

function expectToFailSilently() {
  expect(baseGitProvider.createPrComment).toBeCalledTimes(0)
  expect(baseGitProvider.updatePrComment).toBeCalledTimes(0)
  expect(mockJobRunner.setFailed).toBeCalledTimes(0)
}
