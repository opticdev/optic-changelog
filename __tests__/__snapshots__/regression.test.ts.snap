// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regression shouldnt error on unrelated history 1`] = `
Array [
  Array [
    "{
    \\"errors\\": [
        {
            \\"message\\": \\"Cannot read property 'result' of null\\",
            \\"locations\\": [
                {
                    \\"line\\": 2,
                    \\"column\\": 7
                }
            ],
            \\"path\\": [
                \\"endpointChanges\\"
            ]
        }
    ],
    \\"data\\": {
        \\"endpointChanges\\": null
    }
}",
  ],
  Array [
    "Created body for comment",
  ],
  Array [
    "![changelog](https://raw.github.com/opticdev/optic/develop/assets/changelog_header.png)

[Click Here to See the Documentation](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95)

##### Changelog for  \`/.optic/api/specification.yml\`

###### ðŸŸ¢â€ƒAdded Endpoints (7)

| Endpoint | Description |     |
| -------- | ----------- | --: |
| **POST** /api/v2/person/{person-id}/commands/create_token | Create Token | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_J_IT0RdQice/methods/POST) |
| **GET** /api/account/tokens/latest | DELETE ME | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_kGFkuE-VswW/methods/GET) |
| **GET** /api/account/tokens | Get My Tokens | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_GuLaSKAmMZR/methods/GET) |
| **POST** /api/account/specs | Create Spec Upload | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_I64-7H5OMNT/methods/POST) |
| **GET** /api/account | Get My Account | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_ZOxIrE6uhcD/methods/GET) |
| **GET** /api/specs/{spec-id} | Get Spec by ID | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_4zfSMEGP5g9/methods/GET) |
| **GET** /api/specs/{person-id}/{spec-id} | Get Spec by Person and ID | [**Review**](https://spec.useoptic.com/person/person-id/public-specs/spec-id/changes-since/e037b8d5-3aca-4f6a-b01f-367b00a55c95/paths/path_Y1z6BRzyhv0/methods/GET) |


> Powered by [Optic](https://www.useoptic.com). [Learn how it works!](https://www.useoptic.com/docs/using/baseline)


<!-- optic = {\\"messageHash\\":\\"742463fae01ceca4284d3bde21b63e4a3408e0d3\\"} -->",
  ],
  Array [
    "Creating comment for PR 100",
  ],
]
`;

exports[`Regression shouldnt error on unrelated history 2`] = `
Array [
  Array [
    "Errors found, potentially unrelated base specification:

Cannot read property 'result' of null

GraphQL request:2:7
1 | {
2 |       endpointChanges(sinceBatchCommitId: \\"e037b8d5-3aca-4f6a-b01f-367b00a55c95\\") {
  |       ^
3 |         endpoints {",
  ],
  Array [
    "Assuming regenerated specification",
  ],
]
`;
